client/finality-grandpa/src/finality_proof.rs

```diff
@@ -564,31 +566,32 @@ impl<Header: HeaderT> WarpSyncFragmentCache<Header> {
 		at: Header::Number,
 		item: Option<(AuthoritySetProofFragment<Header>, Header::Number)>,
 	) {
-		if self.cache.len() == self.limit {
-			self.pop_one();
-		}
-		if item.is_some() {
-			// we do not check previous value as cached value is always supposed to
-			// be queried before calling 'new_item'.
-			self.headers_with_justification += 1;
+		self.header_has_proof_fragment.insert(at, item.is_some());
+
+		if let Some(item) = item {
+			if self.cache.len() == self.limit {
+				self.pop_one();
+			}
+
+			self.cache.insert(at, item);
 		}
-		self.cache.insert(at, item);
 	}
 
 	fn pop_one(&mut self) {
-		while let Some(v) = self.cache.pop_front() {
-			if v.1.is_some() {
-				self.headers_with_justification -= 1;
-				break;
-			}
+		if let Some((header_number, _)) = self.cache.pop_front() {
+			self.header_has_proof_fragment.remove(&header_number);
 		}
 	}
 
 	fn get_item(
 		&mut self,
 		block: Header::Number,
-	) -> Option<&mut Option<(AuthoritySetProofFragment<Header>, Header::Number)>> {
-		self.cache.get_refresh(&block)
+	) -> Option<Option<(AuthoritySetProofFragment<Header>, Header::Number)>> {
+		match self.header_has_proof_fragment.get(&block) {
+			Some(true) => Some(self.cache.get_refresh(&block).cloned()),
```

Comment:
I think there is a small issue here:
if I got a header that was removed from cache, since the 'header_has_proof_fragment' is still set to true, then we will return 'Some(None)' which is incorrect in this case.
So if 'get_refresh' return 'Some' we want to retur 'Some(Some' but if it return 'None' we want to just return 'None'.

Reply:
I don't think that this is true, because if we remove the header from the cache in `pop_one()`, we also remove it from `header_has_proof_fragment`.

Reply:
Oh yes, I was thinking pop_one was the lru function :)